### ISSUE TITLE
Publish CLI in a separate Docker image

### ISSUE THREAD

**Issue by nineinchnick** [trinodb/trino#9772]
I'd like to suggest publishing a separate Docker image with only the CLI executable JAR inside. Such an image would be much smaller than the regular Trino image. It might be easier to download and run than the standalone JAR because the following commands should be easy to remember for anyone using Docker on a regular basis:
```sh
docker pull trinodb/trino-cli:363
docker run --rm -it trinodb/trino-cli --server some-remote-cluster --user trino
```
Notice that the entrypoint would be set to trino-cli, making the `docker run` command even shorter.

Right now it's possible with the regular image, but it requires pulling almost 1Gb of data, which can take several minutes or more. Users might not have the image available locally, if they only deploy Trino on remote servers.

**Comment by hashhar** [comment]
I agree with the suggestion. Any concerns @electrum / @mmartint regarding the effect on release process this might have?

**Comment by findepi** [comment]
@hashhar release process includes invoking https://github.com/trinodb/release-scripts/blob/master/trino-docker.sh
so this script should be amended.

**Comment by nineinchnick** [comment]
Note: remember to update the CLI usage section in the description in Docker Hub: https://hub.docker.com/r/trinodb/trino

**Comment by martint** [comment]
The downside of a docker-based CLI (vs downloading the CLI jar file) is that you have to specify which server to connect to. If you’re trying to connect to a local server, you have to jump through some hoops, as “localhost” points at the container itself.

**Comment by nineinchnick** [comment]
It's not meant to replace the downloadable JAR or the CLI bundled in the Trino image. If you have Trino already running and you can `exec` into that container, there's no change there. But if you want to connect to a Trino server running on another host, you might not want to download a 1GB image. Another use case is to run batch jobs from a container (in Kubernetes).

**Comment by martint** [comment]
@nineinchnick , have you seen issues with unicode characters being mangled when running the CLI within docker? 

![image](https://user-images.githubusercontent.com/14387/145247000-9a66a41e-7833-4227-b0e8-8925b5154f0a.png)

**Comment by martint** [comment]
This is why. Not sure why it's picking up that encoding.

```
> docker run -it --rm trinodb/trino:latest java -XshowSettings 2>&1 | grep file.encoding
    file.encoding = ANSI_X3.4-1968
```

**Comment by nineinchnick** [comment]
Looks correct on my laptop:
```
% docker run -it --rm trinodb/trino:latest java -XshowSettings 2>&1 | grep file.encoding
    file.encoding = UTF-8
```

**Comment by martint** [comment]
It seems to be an issue with the arm64 image:

```
❯ docker run -it --platform linux/arm64 --rm trinodb/trino:latest java -XshowSettings 2>&1 | grep file.encoding
    file.encoding = ANSI_X3.4-1968
❯ docker run -it --platform linux/amd64 --rm trinodb/trino:latest java -XshowSettings 2>&1 | grep file.encoding
    file.encoding = UTF-8
```

**Comment by findepi** [comment]
The dockerized CLI is not going to be replacement for "normal" command line usage. (a) `localhost:8080` in-accessibility and (b) lack of history file are going to be sufficient obstacles. Both workaroundable (port forwarding and mounts), but why would I want to do that?
Still, there is use for dockerized CLI as @nineinchnick pointed out. Even if there are encoding issues, we should consider them orthogonal to this. Let's have separate issue about them.

**Comment by martint** [comment]
Yes, I agree. Sorry for hijacking the discussion - I’m not against having a separate image for the CLI. 

Let’s track those other issues separately as they are the kinds of things that would make using such an image less appealing.

**Comment by nineinchnick** [comment]
I started working on this in #10415 but I realized that this new Docker image still has to include the JDK, so it won't that small. As an experiment, I built a truly standalone executable as a GraalVM native image, that is 41MB and doesn't require the JDK to run, only libc. After putting it into a CentOS 8 image, the final compressed size is 97MB, vs 273MB when using the regular JAR with JDK.

GraalVM allows building a fully static binary using muslc, so we could put a single binary into an otherwise empty image. This also works on ARM64. Should I pursue this further?

**Comment by hashhar** [comment]
See older efforts by @wendigo in https://github.com/trinodb/trino/pull/3072

**Comment by nineinchnick** [comment]
@hashhar thanks! Fortunately, the scope here is smaller, as we don't need to build and publish the executable for 3 different OSes using Maven, just for that one Docker image. Also, looks like it's much easier with latest GraalVM 21.3, the resources can be generated automatically using an annotation processor. I'll try building a fully static binary in a Docker image based on scratch.

**Comment by nineinchnick** [comment]
I've managed to build a working static CLI binary. I've pushed the changes to a separate [branch](https://github.com/nineinchnick/trino/compare/trino-cli-image...nineinchnick:trino-cli-image-static) since there are some minor issues: 
* since I used a scratch base image, there's no less - need to find a static replacement or use alpine linux as base
* there are some issues with jline, see below

```
% docker run -it --rm trino:368-SNAPSHOT-cli-amd64 --server 172.17.0.2:8080 
Jan 04, 2022 9:55:07 AM org.jline.utils.Log logr
WARNING: Unable to create a system terminal, creating a dumb terminal (enable debug logging for more information)
trino> show catalogs;
ERROR: failed to open pager: Cannot run program "less": error=2, No such file or directory
 Catalog 
---------
 github  
 hive    
 jmx     
 memory  
 rds     
 system  
 tpcds   
 tpch    
(8 rows)

Query 20220104_095642_00008_6hj3f, FINISHED, 1 node
Splits: 19 total, 19 done (100.00%)
0.07 [0 rows, 0B] [0 rows/s, 0B/s]

trino>             
```

The image is tiny:
```
% docker images | grep SNAPSHOT
trino                                                            368-SNAPSHOT-cli-amd64         8522cfa10546   2 minutes ago       51.1MB
```
and should only require to download about 17MB:
```
% docker save trino:368-SNAPSHOT-cli-amd64 -o trino-cli.tar
% gzip trino-cli.tar 
% du -ms trino-cli.tar.gz
17	trino-cli.tar.gz
```

I wanted to do a sanity check again - should I continue with this idea and iron out the remaining issues?

CC @electrum since you reviewed the PR but you're not yet subscribed to this issue.

**Comment by nineinchnick** [comment]
I opened #14852 so we can have some discussion there if/how to proceed.
